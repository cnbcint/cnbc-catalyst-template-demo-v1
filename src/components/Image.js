import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { breakpoints } from "../helpers/Layout";
import useAnimation from "../hooks/useAnimation";
const ImageElement = styled.img`
  width: 100%;
`;
/**
Responsive image component with lazyload tag. 
This components required images for desktop tablet and mobile screensizes. 
Each image can have it's own aspect ratio to achive more flexibility on design.
The image will take 100% of space available from parent container.
 */
export default function Image(props) {
  const { ref: refImage } = useAnimation({
    animationName: "fadeIn",
    duration: 500,
    delay: 0,
    loop: true,
    autoStart: true,
  });

  return (
    <picture className={props.className}>
      <source
        media={`(min-width:${breakpoints.tablet})`}
        srcSet={props.source.desktop}
      />
      <source
        media={`(min-width:${breakpoints.mobile})`}
        srcSet={props.source.tablet}
      />
      <ImageElement
        ref={refImage}
        src={props.source.mobile}
        style={props.style}
        alt={props.alt}
        loading="lazy"
      />
    </picture>
  );
}

Image.propTypes = {
  className: PropTypes.string,
  alt: PropTypes.string.isRequired,
  /**
   Image url source for desktop tablet and mobile
   */
  source: PropTypes.shape({
    mobile: PropTypes.string,
    tablet: PropTypes.string,
    desktop: PropTypes.string,
  }).isRequired,
  /**
   * React CSS overrides to IMG element
   */
  style: PropTypes.object,
};

Image.defaultProps = {
  className: "",
};
