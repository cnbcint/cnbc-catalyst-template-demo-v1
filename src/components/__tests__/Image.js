import React from "react";
import Image from "../Image";
import { render } from "@testing-library/react";
import "jest-styled-components";

const testProps = {
  source: {
    desktop:
      "http://localhost:8080/images/responsive/output/image05-desktop.jpg",
    tablet: "http://localhost:8080/images/responsive/output/image05-tablet.jpg",
    mobile: "http://localhost:8080/images/responsive/output/image05-mobile.jpg",
  },
  alt: "Alt Text Test",
};

beforeAll(() => {
  window.IntersectionObserver = class {
    observe() {
      return null;
    }
    disconnect() {}
  };
});

describe("Image", () => {
  it("Renders correctly", () => {
    const { container } = render(<Image {...testProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Renders an img with default mobile source", () => {
    const { container } = render(<Image {...testProps} />);
    expect(container.querySelector("img").src).toMatch(
      "http://localhost:8080/images/responsive/output/image05-mobile.jpg"
    );
  });

  it("Renders img alt text", () => {
    const { container } = render(<Image {...testProps} />);
    expect(container.querySelector("img").alt).toMatch("Alt Text Test");
  });

  it("image is lazy load enable", () => {
    const { container } = render(<Image {...testProps} />);
    expect(container.querySelector("img").getAttribute("loading")).toMatch(
      "lazy"
    );
  });

  it("Add sources for desktop and tablet", () => {
    const { container } = render(<Image {...testProps} />);
    expect(container.querySelectorAll("source")[0].srcset).toMatch(
      "http://localhost:8080/images/responsive/output/image05-desktop.jpg"
    );
    expect(container.querySelectorAll("source")[1].srcset).toMatch(
      "http://localhost:8080/images/responsive/output/image05-tablet.jpg"
    );
  });

  it("Sources has the correct media queries for tablet and desktop", () => {
    const { container } = render(<Image {...testProps} />);
    expect(container.querySelectorAll("source")[0].media).toMatch(
      "(min-width:1280px)"
    );
    expect(container.querySelectorAll("source")[1].media).toMatch(
      "(min-width:768px)"
    );
  });
});

afterAll(() => {
  window.IntersectionObserver = null;
});
