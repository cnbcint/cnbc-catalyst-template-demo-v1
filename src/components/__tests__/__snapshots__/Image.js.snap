// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Renders correctly 1`] = `
.c0 {
  width: 100%;
}

<picture
  class=""
>
  <source
    media="(min-width:1280px)"
    srcset="http://localhost:8080/images/responsive/output/image05-desktop.jpg"
  />
  <source
    media="(min-width:768px)"
    srcset="http://localhost:8080/images/responsive/output/image05-tablet.jpg"
  />
  <img
    alt="Alt Text Test"
    class="c0 pre-animated"
    loading="lazy"
    src="http://localhost:8080/images/responsive/output/image05-mobile.jpg"
  />
</picture>
`;
